File: Code/Ch1-memory-centric-system-model/cpp-code/abstract_system_peripherals.cc
#include <vector>
#include <tuple>
//## Peripheral actions
#define True 1
#define False 0

std::tuple<std::vector<uint64_t>,uint64_t> kbdAction(std::vector<uint64_t>& kbdState){
    uint64_t kbdIrq;
    uint64_t None=0;
    kbdIrq=False;
    return std::make_tuple(kbdState,kbdIrq);

}//4 > 0

std::tuple<std::vector<uint64_t>,uint64_t> nicAction(std::vector<uint64_t>& nicState){
    uint64_t nicIrq;
    uint64_t None=0;
    nicIrq=False;
    return std::make_tuple(nicState,nicIrq);

}//4 > 0

std::tuple<std::vector<uint64_t>,uint64_t> ssdAction(std::vector<uint64_t>& ssdState){
    uint64_t ssdIrq;
    uint64_t None=0;
    ssdIrq=False;
    return std::make_tuple(ssdState,ssdIrq);

}//4 > 0

std::tuple<std::vector<uint64_t>,uint64_t> gpuAction(std::vector<uint64_t>& gpuState){
    uint64_t gpuIrq;
    uint64_t None=0;
    gpuIrq=False;
    return std::make_tuple(gpuState,gpuIrq);

}//4 > 0

std::tuple<std::vector<uint64_t>,uint64_t> timerAction(std::vector<uint64_t>& timerState){
    uint64_t timerIrq;
    uint64_t None=0;
    if (timerState[0]>0 ) {
        timerState[0]-=1;
    }//8 > 4
    timerIrq=False;
    if (timerState[0]==0 ) {
        timerIrq=True;
        timerState[0]=timerState[1];
    }//8 > 4
    return std::make_tuple(timerState,timerIrq);
}

File: Code/Ch1-memory-centric-system-model/advanced-model/abstract_system_peripherals.py
from abstract_system_constants import *

### Peripheral actions

def kbdAction(kbdState):
    kbdIrq=False
    return (kbdState,kbdIrq)        

def nicAction(nicState):
    nicIrq=False
    return (nicState,nicIrq)  

def ssdAction(ssdState):
    ssdIrq=False
    return (ssdState,ssdIrq)  

def gpuAction(gpuState):
    gpuIrq=False
    return (gpuState,gpuIrq)  

def timerAction(timerState):
    # print('timerAction',timerState)
    if timerState[0]>0:
        timerState[0]-=1
    timerIrq=False
    if timerState[0]==0:
        # print('timerAction: timerIrq SET')
        timerIrq=True
        timerState[0]=timerState[1]
    return (timerState,timerIrq)

def dmaAction(systemState):
    dmaIrq=0
    # DMA is the start of the address space
    # DCR values: 1 = do transfer, 0 = idle
    if  systemState[DMA+DCR]!=0:
        if systemState[DMA+DCO]!=0:
            ctr = systemState[DMA+DCO]
            to_addr = systemState[DMA+DDR]+ctr
            from_addr = systemState[DMA+DSR]+ctr             
            systemState[to_addr] = systemState[from_addr]
            systemState[DMA+DCO]=-1
        systemState[DMA+DCR]=0
        dmaIrq=1
    return (systemState,dmaIrq)
